### PR TITLE
feature(geocoder): Add support for HERE Geocoder

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
This commit adds support for HERE Geocoder to the geocoder package.

### CODE REVIEW

**Review by miles-grant-ibigroup — COMMENTED** [review]
Great work! This looks really great. I don't know if mocking is enabled for the geocoder stories, but they don't do anything for me at the moment.

Tested this in a full OTP deployment and it works perfectly (even our gray location details)! Thank you so much.

I just have a few tiny nitpicks!

File: packages/location-field/src/stories/common.js [hunk 1]
```diff
@@ -17,6 +17,13 @@ export const geocoderConfig = {
   maxNearbyStops: 4,
   type: "PELIAS"
 };
+
+export const hereGeocoderConfig = {
+  type: "HERE",
+  apiKey: "fo_vzDfPGAYXJTG3HbdEZutV-4vjsP0LyekRDac_XaY",
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/location-field/src/stories/common.js:23]: No. Key shouldn't be in here of course, but it's just from an account I made for testing.

**Review by miles-grant-ibigroup — APPROVED** [review]
This looks really great! Thanks so much really excited about this

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  ismorphicMapzenSearchQuery?: string,
+  lat: number,
+  lon: number,
+  name: string,
+  rawGeocodedFeature: Feature
+}
+/**
+ * Create customized geocoder functions given a certain geocoding API, the
+ * config for the geocoder and response rewrite functions specific to this
+ * application. Any geocoder api that is added is expected to have an API that
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:78]: api -> API

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+  api: GeocoderAPI
+
+  constructor(geocoderApi: GeocoderAPI, geocoderConfig: GeocoderConfig) {
+    this.api = geocoderApi;
+    this.geocoderConfig = geocoderConfig;
+  }
+
+  /**
+   * Perform an autocomplete query. Eg, using partial text of a possible
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:92]:
```suggestion
   * Perform an autocomplete query, e.g. using partial text of a possible
```

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        name: feature.properties.label,
+        rawGeocodedFeature: feature
+      } 
+      return Promise.resolve(location);
+    }
+    return Promise.reject(new Error("Feature is not of type Point."))
+  }
+
+  /**
+   * Do a reverse-geocode. ie get address information and attributes given a
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:123]:
```suggestion
   * Do a reverse-geocode, i.e. get address information and attributes given a
```

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  reverse(query: ReverseQuery): Promise<SingleGeocoderResponse> {
+    return this.api
+      .reverse(this.getReverseQuery(query))
+      .then(this.rewriteReverseResponse);
+  }
+
+  /**
+   * Perform a search query. A search query is different from autocomplete in
+   * that it is assumed that the text provided is more or less a complete
+   * well-fromatted address.
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:135]: Typo f**or**matted

File: packages/geocoder/src/geocoders/here.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import type { Feature } from "geojson"
+import type { HereResponse, Item } from "../apis/here/types";
+import type { SingleGeocoderResponse, MultiGeocoderResponse } from "./abstract-geocoder";
+
+import Geocoder from "./abstract-geocoder";
+
+
+const hereResultTypeToPeliasLayerMap = {
+  place: "venue",
+  houseNumber: "address"
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/here.ts:12]: Sort attributes (multiple instances in this file).

File: packages/geocoder/src/geocoders/here.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
+// Prettier does not support typescript annotation
+// eslint-disable-next-line prettier/prettier
+import type { Feature } from "geojson"
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/here.ts:3]: Please lint this file (see my comment in `here.ts`).

File: packages/geocoder/src/geocoders/here.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+  return {
+    type: "Feature",
+    geometry: {
+      type: "Point",
+      coordinates: [hereFeature.position.lng, hereFeature.position.lat]
+    },
+    properties: {
+      source: "here",
+      label: hereFeature.address.label,
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/here.ts:34]: (I'm tempted to create a var for `hereFeature.address`.)

File: packages/geocoder/src/index.story.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
+import React, { useState } from "react";
+import PropTypes from "prop-types";
+import getGeocoder from "./index";
+
+const GeocoderTester = ({
+  hereApiKey,
+  geocodeEarthApiKey,
+  onResults,
+  endpoint,
+  at
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.js:10]: Sort props and attributes (multiple instances in this file).

File: packages/geocoder/src/index.story.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    const peliasRes = enableGeocodeEarth
+      ? await peliasGeocoder[endpoint]({
+          text: searchTerm,
+          focusPoint,
+          point: focusPoint
+        })
+      : {};
+    onResults({
+      hereRes,
+      peliasRes
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.js:48]: Just making sure, is the return value `{}` going to be correctly handled when otherwise a promise is returned?

File: packages/geocoder/src/index.story.js [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+      <div>
+        <label htmlFor="lon">
+          lng: lat:
+          <input
+            type="text"
+            id="lon"
+            onChange={e => {
+              setFocusPoint({ ...focusPoint, lat: e.target.value });
+            }}
+            value={focusPoint.lat}
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.js:63]: Sort props (multiple instances in this file).

File: packages/geocoder/src/index.story.js [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+    lng: PropTypes.number
+  })
+};
+
+GeocoderTester.defaultProps = {
+  at: {
+    lat: 0,
+    lng: 0
+  }
+};
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.js:136]: This file should be a `tsx` file, in which these would become types.

File: packages/geocoder/src/index.ts [hunk 13]
```diff
@@ -17,10 +23,13 @@ const getGeocoder = memoize(geocoderConfig => {
...
+    case "HERE":
+      return new HereGeocoder(here, geocoderConfig);
     default:
       console.error(`Unkown geocoder type: "${type}". Using NoApiGeocoder.`);
       return new NoApiGeocoder();
   }
 });
 
 export default getGeocoder;
+export type { SearchQuery, AutocompleteQuery, ReverseQuery };
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.ts:35]: Add empty line at end of file.

File: packages/geocoder/src/index.test.js [hunk 14]
```diff
@@ -19,6 +19,10 @@ describe("geocoder", () => {
       baseUrl: "https://ws-st.trimet.org/pelias/v1",
       type: "PELIAS"
     },
+    {
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.test.js:22]: This file should have `ts` extension too.

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
It sounds from other comments that the functionality has been tested, but I have flagged quite a few formatting issues.

File: packages/geocoder/src/apis/here/index.ts [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+  apiKey: string
+}
+
+// These types are standardized for the other geocoders in this library.
+// Perhaps we could extract them out somewhere and reuse them in the other libraries? 
+type HereFetchArgs = {
+  options: RequestInit // Built-in Typing
+  query: HereQuery
+  url: string
+}
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/apis/here/index.ts:31]: I was confused about the same thing! I I still am not sure why disable-next-line would break linting for the whole file. Is there a newer version of prettier that can handle `import type`?

[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/apis/here/index.ts:31]: Probably, but someone would have to write a PR so that only modified files are linted, to avoid failing CI jobs unnecessarily.

File: packages/geocoder/src/geocoders/here.ts (hunk 8 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/here.ts:33]: extracted a bunch of things, looks way better!

File: packages/geocoder/src/index.test.js (hunk 14 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/index.test.js:22]: Making this file typed would create a lot of work to fix all the mocks and other things to use correct data structures when it's not necessary. Is there a better way of typing a file like this?

File: packages/geocoder/src/index.story.js (hunk 10 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/index.story.js]: Yes, it's not a problem for Storybook.

File: packages/geocoder/src/index.test.js (hunk 14 above)
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/index.test.js:22]: Does just changing the extension to `.ts` cause TypeScript errors? (TypeScript can infer types and will enforce only where other code requires them).

File: packages/geocoder/src/apis/here/index.ts [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+ * @param  {string} $0.text                       query text
+ * @return {Promise}                              A Promise that'll get resolved with the autocomplete result
+ */
+function autocomplete({
+  apiKey,
+  focusPoint,
+  boundary,
+  size = 20,
+  options,
+  text
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/apis/here/index.ts]: `boundary` is out of order.

File: packages/geocoder/src/apis/here/index.ts [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+}: AutocompleteQuery): JSONArrayPromise {
+  // build query
+  const query: HereQuery = { apiKey, q: text, limit: size, show: "details" }
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/apis/here/index.ts]: `q` is out of order.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 18]
```diff
@@ -0,0 +1,223 @@
...
+
+type Boundary = {
+  country?: string;
+  rect?: Rect;
+};
+
+type GeocoderAPI = {
+  autocomplete: (query: AutocompleteQuery) => Promise<Array<JSON>>;
+  search: (query: SearchQuery) => Promise<Array<JSON>>;
+  reverse: (query: ReverseQuery) => Promise<Array<JSON>>;
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:23]: Sort attributes.

File: packages/geocoder/src/geocoders/here.ts [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+    };
+  }
+
+  return {
+    type: "Feature",
+    geometry: {
+      type: "Point",
+      coordinates: [position.lng, position.lat]
+    },
+    properties: {
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/here.ts:32]: Sort attributes (multiple instances in this file)

File: packages/geocoder/src/index.story.tsx [hunk 20]
```diff
@@ -0,0 +1,163 @@
...
+    size: 1,
+    type: "PELIAS"
+  });
+
+  const search = async () => {
+    const hereRes = enableHere
+      ? await geocoder[endpoint]({
+          text: searchTerm,
+          focusPoint,
+          point: focusPoint
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.tsx:41]: Sort attributes

File: packages/geocoder/src/index.story.tsx [hunk 21]
```diff
@@ -0,0 +1,163 @@
...
+  title: "Geocoder",
+  component: (
+    <GeocoderTester
+      endpoint="search"
+      geocodeEarthApiKey="placeholder"
+      hereApiKey="placeholder"
+      onResults={() => null}
+    />
+  ),
+  argTypes: { onResults: { action: "result" } }
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.tsx:135]: Sort attributes (multiple instances in this file).

**Review by binh-dam-ibigroup — CHANGES_REQUESTED** [review]
After discussion with @miles-grant-ibigroup, since you started adding TypeScript to this package, we felt you should also convert `pelias.js` and `noapi.js` to TypeScript, please. Also see other minor formatting comments.

File: packages/geocoder/src/geocoders/pelias.ts [hunk 22]
```diff
@@ -11,7 +14,7 @@ export default class PeliasGeocoder extends Geocoder {
    * Generate an autocomplete query specifically for the Pelias API. The
    * `sources` parameter is a Pelias-specific option.
    */
-  getAutocompleteQuery(query) {
+  getAutocompleteQuery(query: AutocompleteQuery): AutocompleteQuery {
```
Comment (miles-grant-ibigroup) [inline packages/geocoder/src/geocoders/pelias.ts:17]: It feels a bit strange to have the input and output here be the same type. I would maybe again add a comment as to what's going on, even if the types are correct.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 23]
```diff
@@ -0,0 +1,231 @@
...
+  format?: boolean;
+  options?: RequestInit;
+  size?: number;
+  text?: string;
+  url?: string;
+  sources?: string;
+};
+
+export type MultiGeocoderResponse = FeatureCollection & {
+  isomorphicMapzenSearchQuery?: SearchQuery & AutocompleteQuery & ReverseQuery;
```
Comment (miles-grant-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:72]: Should this be a `|` instead of `&`?

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 24]
```diff
@@ -0,0 +1,231 @@
...
+ */
+export default class Geocoder {
+  geocoderConfig: GeocoderConfig;
+
+  api: GeocoderAPI;
+
+  constructor(geocoderApi?: GeocoderAPI, geocoderConfig?: GeocoderConfig) {
+    this.api = geocoderApi;
+    this.geocoderConfig = geocoderConfig;
+    console.log(this.geocoderConfig);
```
Comment (miles-grant-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:96]: Uh oh I think this snuck in by accident!

File: packages/geocoder/src/index.test.js [hunk 25]
```diff
@@ -33,6 +37,7 @@ describe("geocoder", () => {
     .replyWithFile(200, mockResponsePath("arcgis", "suggest-response.json"))
     // reverse
     .get(`${baseArcGisPath}reverseGeocode`)
+    .twice()
```
Comment (miles-grant-ibigroup) [inline packages/geocoder/src/index.test.js:40]: Didn't know about this but it's cool! Nice one

File: packages/location-field/src/mocks/handlers.js [hunk 26]
```diff
@@ -1,12 +1,19 @@
 import { rest } from "msw";
```
Comment (miles-grant-ibigroup) [inline packages/location-field/src/mocks/handlers.js:1]: Did you add this handler to `.storybook/preview.js`? I'm only getting an empty object back when I run the stories locally. I also don't get anything back at all for the reverse story.

https://github.com/opentripplanner/otp-ui/pull/328/files also adds a new msw endpoint

File: packages/geocoder/src/index.story.tsx [hunk 27]
```diff
@@ -0,0 +1,183 @@
...
+        })
+      : {};
+    onResults({
+      hereRes,
+      peliasRes
+    });
+  };
+
+  return (
+    <div>
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/index.story.tsx:64]: It's to separate the next field onto a new line. Would a <br/> be better?

File: packages/location-field/src/mocks/handlers.js (hunk 26 above)
Comment (daniel-heppner-ibigroup) [reply packages/location-field/src/mocks/handlers.js:1]: Oh, I was testing the story by putting real API keys into the story controls. The use-case for the story, for me anyway, was to look at and compare the responses from HERE and Pelias for various searches, so mocking the responses didn't really make sense. At this stage, though, it might make more sense, so I'll look into adding that. You're probably getting weird responses because the API keys are not correct.  

I also haven't touched `preview.js` before, and I'm not sure what it does. I'll take a look and see if I can't figure that out!

File: packages/geocoder/src/geocoders/here.ts [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+    type: "Feature"
+  };
+};
+
+export default class HereGeocoder extends Geocoder {
+  rewriteReverseResponse({ items, point }: HereResponse): MultiGeocoderResponse | SingleGeocoderResponse {
+    if (this.geocoderConfig?.reverseUseFeatureCollection) {
+      return {
+        features: items
+        .map(convertHereToGeojson),
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/here.ts:59]: I probably had more stuff in there before I refactored convertHereToGeojson into a function, and this is an artifact of that. good catch

File: packages/geocoder/src/index.test.js (hunk 25 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/index.test.js:40]: 😁

File: packages/geocoder/src/geocoders/abstract-geocoder.ts (hunk 23 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts:72]: I thought so too, but no that doesn't work. A `|` creates a type that needs to conform to any of the types, so it expects fields from all of them. Seemed counterintuitive to me, but this is what ended up working.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 29]
```diff
@@ -0,0 +1,231 @@
...
+  sources?: string;
+  size?: number;
+  reverseUseFeatureCollection?: boolean;
+};
+
+export type ReverseQuery = {
+  apiKey?: string;
+  format?: boolean;
+  options?: RequestInit;
+  point?: LonLatInput;
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts:42]: There's no types.ts at this level (there is one in `here`), but there probably should be.

File: packages/geocoder/src/geocoders/here.ts (hunk 28 above)
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/here.ts]: Fixed in #a3cc58b

File: packages/geocoder/src/geocoders/abstract-geocoder.ts (hunk 24 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts]: Fixed in a3cc58b

File: packages/geocoder/src/geocoders/pelias.ts (hunk 22 above)
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/pelias.ts:19]: Commented in a3cc58b

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 30]
```diff
@@ -0,0 +1,231 @@
...
+      url: baseUrl ? `${baseUrl}/search` : undefined,
+      format: false, // keep as returned GeoJSON,
+      ...query
+    };
+  }
+
+  /**
+   * Default rewriter for autocomplete responses
+   */
+  rewriteAutocompleteResponse(response: unknown): MultiGeocoderResponse {
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts:214]: Commented in a3cc58b

File: packages/location-field/src/mocks/handlers.js (hunk 26 above)
Comment (daniel-heppner-ibigroup) [reply packages/location-field/src/mocks/handlers.js:1]: Nevermind my earlier comment. I didn't realize you were talking about location-field.

Comment (daniel-heppner-ibigroup) [reply packages/location-field/src/mocks/handlers.js:1]: A little confused here. The location-field is working fine for me, and the handler for it *is* making it into preview.js via the array that is passed in. I guessing that you are actually talking about the Geocoder package, in which case my original comment applies. I'll go ahead and create an MSW mock for it.

Comment (daniel-heppner-ibigroup) [reply packages/location-field/src/mocks/handlers.js:1]: OK, I believe I've taken care of what you're asking for in e4def13. Let me know if that works!

File: packages/geocoder/src/index.story.tsx (hunk 27 above)
[the remark this answers is not in the document]
Comment (miles-grant-ibigroup) [reply packages/geocoder/src/index.story.tsx:64]: I think CSS might be the way to go but I think for this story we should be fine. Will leave it up to Binh to decide

File: packages/location-field/src/mocks/handlers.js (hunk 26 above)
Comment (miles-grant-ibigroup) [reply packages/location-field/src/mocks/handlers.js:1]: This was a journey! But yes I was talking about creating mocks for the geocoder story and this looks excellent and works great thank you so much for implementing!

**Review by miles-grant-ibigroup — APPROVED** [review]
In my eyes this is good to go! This turned out really excellent thanks for all your work on getting this just right. 

Have tested this with otp-rr and we can swap HERE and pelias with just 2 config line changes! It could not have gone any better. Thanks again!

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 31]
```diff
@@ -18,17 +18,21 @@ type Boundary = {
...
   baseUrl?: string;
   boundary?: Boundary;
   focusPoint?: LonLatInput;
   options?: RequestInit;
+  sources?: string;
+  size?: number;
+  reverseUseFeatureCollection?: boolean;
 };
 
 export type ReverseQuery = {
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:38]: Any thoughts @daniel-heppner-ibigroup @miles-grant-ibigroup about creating an interface with some of the fields common to all query types below? (this is not blocking)

File: packages/geocoder/src/geocoders/abstract-geocoder.ts (hunk 30 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts:214]: Add that comment to the fields of type `Promise<unknown>`.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 32]
```diff
@@ -202,20 +209,23 @@ export default class Geocoder {
...
     return response as MultiGeocoderResponse;
   }
 
   /**
    * Default rewriter for reverse responses
+   * Response type is unknown because it depends on the specific Geocoder implementation.
    */
-  rewriteReverseResponse(response: unknown): MultiGeocoderResponse {
+  rewriteReverseResponse(response: unknown): MultiGeocoderResponse | SingleGeocoderResponse {
     return response as MultiGeocoderResponse;
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:223]: The `return` statement type contradicts the declared return type.

File: packages/geocoder/src/geocoders/here.ts [hunk 33]
```diff
@@ -37,8 +36,8 @@ const convertHereToGeojson = (hereFeature: Item): Feature => {
       housenumber: address.houseNumber,
       label: address.label,
       layer: hereResultTypeToPeliasLayerMap[resultType]
-        ? hereResultTypeToPeliasLayerMap[resultType]
-        : resultType,
+      ? hereResultTypeToPeliasLayerMap[resultType]
+      : resultType,
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/here.ts:40]: Revert the change in indentation.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts [hunk 34]
```diff
@@ -125,10 +132,10 @@ export default class Geocoder {
    * Do a reverse-geocode, i.e. get address information and attributes given a
    * GPS coordinate.
    */
-  reverse(query: ReverseQuery): Promise<MultiGeocoderResponse> {
+  reverse(query: ReverseQuery): Promise<MultiGeocoderResponse | SingleGeocoderResponse> {
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/geocoders/abstract-geocoder.ts:135]: Maybe create an alias `type SingleOrMultiGeocoderResponse = MultiGeocoderResponse | SingleGeocoderResponse` and reuse? (~5 occurrences)

File: packages/geocoder/src/index.story.tsx [hunk 35]
```diff
@@ -18,34 +18,40 @@ const GeocoderTester = ({
...
+    reverseUseFeatureCollection,
+    setReverseUseFeatureCollection
+  ] = useState(false);
 
   const geocoder = getGeocoder({
     apiKey: hereApiKey,
     focusPoint,
-    type: "HERE"
+    type: "HERE",
+    reverseUseFeatureCollection
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.tsx:30]: Sort attributes (2 instances)

File: packages/geocoder/src/index.story.tsx [hunk 36]
```diff
@@ -95,24 +101,39 @@ const GeocoderTester = ({
...
-        />
-      </label>
+      <div>
+        <label htmlFor="geocodeEarth">
+          Geocode.earth:
+          <input
+            id="geocodeEarth"
+            onChange={e => setEnableGeocodeEarth(e.target.checked)}
+            type="checkbox"
+            checked={enableGeocodeEarth}
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.story.tsx:111]: Sort props (2 instances).

File: packages/geocoder/src/index.ts [hunk 37]
```diff
@@ -10,10 +10,10 @@ import HereGeocoder from "./geocoders/here";
 
 // Prettier does not support typescript annotation
 // eslint-disable-next-line prettier/prettier
-import type { SearchQuery, ReverseQuery, AutocompleteQuery } from "./geocoders/abstract-geocoder"
+import type { SearchQuery, ReverseQuery, AutocompleteQuery, GeocoderConfig } from "./geocoders/abstract-geocoder"
```
Comment (binh-dam-ibigroup) [inline packages/geocoder/src/index.ts:13]: Sort imports.

File: packages/geocoder/src/index.story.tsx (hunk 27 above)
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply packages/geocoder/src/index.story.tsx:64]: That's fine for the story's purposes.

**Review by binh-dam-ibigroup — APPROVED** [review]
LGTM subject to the minor refactoring, formatting, and comment suggestions.

File: packages/geocoder/src/geocoders/abstract-geocoder.ts (hunk 31 above)
Comment (miles-grant-ibigroup) [reply packages/geocoder/src/geocoders/abstract-geocoder.ts:38]: Could be good! Could put it in the types package. But agree not needed for this PR
